New Runtime Implementation

dolio opened the pull request:
This pull request contains the new runtime work. By default, the existing runtime is still used, but the new one can be enabled by passing `--new-runtime` on the command line, or putting `new-runtime: true` in the config file.

The majority of the test suite passes using the new runtime. One failure is an actual incorrect test case (#1613). The other is the recently added IO case, which I haven't looked into just yet. There are also some transcript differences that I haven't resolved yet.

In general, the `IO` implementation is relatively untested, and may have bugs, but I thought this pull request might signal that someone besides me could actually try things out, as I might not find all the issues myself.

The extra levels of code generation down to the executable layer are far from ideal, but even so some simple loops should run faster than the existing runtime. Loops with things like ability usage might be worse or the same, because those are the areas that are noticeably far from optimal with respect to code generation/optimization.

One significant thing I have not done is put the new runtime in a separate subproject, even though I think it should probably be in one. This is because the main program is under `parser-typechecker/`, but that portion of the project also contains a lot of machinery that is prior to the new runtime and code generator layers for it. If the project reorganization is merged prior to this pull request, I can rearrange my new code to be somewhat more self-contained.

Questions/comments welcome.

dolio committed:
Begin implementation of a new runtime.

- Some description of the ideas behind the design
- New low level code representation
- Interpreter for the low level code
- Unboxed data stack based on arrays from the primitive
  package to avoid redundant offset arithmetic
- Benchmark of the new interpreter currently using a simple
  unboxed loop

dolio committed:
Further new runtime progress

- Added a boxed stack (not used much yet)
- Added frame pointers to data stacks for tracking size (trying to
  compute these statically seems like considerable work for questionable
  gain at this point).
- Various changes for having both boxed and unboxed value management
- Instruction changes based on ideas for calling conventions.

dolio committed:
Some code cleanup in preparation for further development.

- Broke out an apply function for proper handling of partial
  application and the like
- Created a class to overload the data stack API
  - Doesn't seem to have a performance cost, as everything
    is used with a particular concrete type. One oddity is
    that `Seg 'BX` cannot be unpacked according to GHC, even
    though what it should reduce to, `Array Closure` can.

dolio committed:
Indent benchmark a bit nicer.

dolio committed:
Flesh out boxed implementation and assorted fixes

- Flesh out stack implementation for capturing and pushing segments
- Implement more boxed instructions in the evaluator
- Add stack consumption information back to App and Yield
- Add boxed loop to benchmark

dolio committed:
Add more specific data closure cases.

- Mirrors Args cases
- Significantly improves boxed loop performance; worker-wrapper
  may no longer be as necessary.

dolio committed:
Merge remote-tracking branch 'origin/master' into wip/rt2

dolio committed:
Remove single box closure constructor.

dolio committed:
Document the constructors of IR and Branch

dolio committed:
Toward implementation of delimited control.

- Still has some issues; the operations do not do quite the
  right thing.

dolio committed:
Benchmark test case using delimited operations

- Implements state using shift/reset and does the loop using
  the state as the accumulator.

dolio committed:
Merge branch 'master' into wip/rt2

dolio committed:
Various reworking and fixing to make continuations work

- Benchmark test case doing a loop with state implemented
  via delimited control works.

dolio committed:
Describe plan for dynamically scoped variables

dolio committed:
Document the meaning of IR constructors more

dolio committed:
Add fast-path application

- Isn't a huge improvement, but it is better

dolio committed:
Add a by-name allocation instruction

dolio committed:
Implement delimited dynamic binding references

- Benchmark test case using local state threading in conjunction with
  the dynamic scoping for a continuation-less effect handler.
- Assorted cleanup as well

dolio committed:
Factor out calculation of pointers for grab

dolio committed:
Fix augSeg joining segments in the wrong order

dolio committed:
Try structuring the abstract machine code more

dolio committed:
Inline some structure for more efficiency

dolio committed:
More performance-oriented restructuring

dolio committed:
Reintroduce Let as part of sections

- This seems like a more natural placement than having a 'push a return
  to the continuation instruction.
- Also add some missing pragmas that seemed to be limiting performance

dolio committed:
Rename IR2 to MCode, since it's not really 'intermediate'

dolio committed:
Start building an ABT for normalized trees

dolio committed:
Implement new ANF transformation

- Uses a new, explicitly normalized representation of terms.
- Also assumes that the terms have already been compiled to
  supercombinators, as there is no representation of local
  lambdas or letrec.

dolio committed:
Merge branch 'master' into wip/rt2

dolio committed:
Fix ANF code wrt renaming of AnnotatedTerm

dolio committed:
Miscelaneous ANF additions

- Added some additional pattern synonyms to cover ANF terms
- Cleaned up some ANF code using the new patterns
- More thorough exporting of new types/functions

dolio committed:
Add some tests for ANF translator

- Currently testing that going back from the ANF translation gives
  an equivalent result for simple examples. This won't necessarily
  be true for every term, though (intentionally so).

dolio committed:
Restructure normalized ABT and ANF code slightly

- Moved variable occurrences out of the ABT structure. Since the
  variable type gets passed directly to the shape functor, it's not
  necessary for the ABT to handle variable occurences, and the
  factorization was interfering with the structure of the algorithm.
- Made the ABT and Term types directly recursive. There doesn't seem to
  be a reason for them not to be
- Factored the ANF shape functor into two types to ensure that there is
  no let binding with a let directly inside it. Instead lets must be
  linearized. This is arguably more normalized, and should correspond
  closer to CPS.

dolio committed:
Implemented a function to 'substitute' into an ANF

- True substitution does not make sense on ANF, but we can insert a
  bindings for a term at the lowest points in the tree as possible.
  The function does share the value in some situations, however.

dolio committed:
Flesh out matching ANF

dolio committed:
Restructured ANF in view of ability compilation

- ANF handle construct takes direct branching, and corresponds to
  surface syntax matching on `Request`. This is the location where
  we know which abilities are in play and can install handlers
  without complicated information passing.
- The ANF handle is distinct from other case analysis because the
  scrutinee need not already be evaluated. One could conceive of
  this as being call-by-name case analysis rather than call-by-value,
  although that is not how it will be implemented.
- This means that a function that does Request matching essentially
  needs to have a call-by-name argument, and the surface-level handle
  construct is a form of call-by-name application. Therefore, it
  compiles into a by-name correspondent to let.
- However, this also probably means that Request and handle are
  inessential in the surface syntax. `Request {e} a` is essentially
  the same as `'{e} a`, handle quotes its body, and ability matching
  forces the suspended computation. If the type checker were modified
  such that matching could provide abilities to its scrutinee, this
  scheme would work with quote and !.

dolio committed:
Some ANF tweaks and most of an ANF->Interpreter phase

- Tweaked ANF by-name bindings to match the fact that machine by-name
  instructions are to known function calls. This is due to delayed code
  needing to access local environment outside of the stack discipline,
  and the way we manage this sort of environment is through
  supercombinators.
- The compiler from ANF to interpreter code is nearly complete (for a
  first pass), except for handler code, which is the trickiest part. The
  difficulty comes from the fact that a 'handle' construct needs to
  become a combinator that responds to effect requests, _plus_ code that
  installs that combinator in the right spot. Someone needs to divide up
  these into separate combinators, and probably the ANF phase is better
  for this.
- Other assorted tweaks as part of this process.

dolio committed:
Monadify ANF code

- Some previously explicit threading is now implicit. Variable
  avoidance is tracked using state now, with an operation for
  resetting this state at opportune points. This may have fixed
  some mistakes in the prior code (and over-avoiding is not much
  of a problem).
- State is also carried for creating multiple supercombinators
  for a single term. This is necessary for creating functions
  that respond to effects and install handlers from a single
  matching term.
- Reference resolution is carried as a reader.

dolio committed:
Finish preliminary compilation of ANF->MCode

- Some MCode has changed in response to the demands of the higher
  level.
  * Reset can now simultaneously delimit a set of prompts,
    and the interpreter has changed to accomodate this.
  * A Die instruction has been added for implementing empty case
    statements.
- Some ANF has been restructured as well. E.G. ability matching
  moved back into the same type as data matching.
- The calling convention of some things might change in the
  future. For instance, currently handling is implemented as a
  single function. The ability request is expected to have a tag
  corresponding to the ability, and an unboxed argument
  corresponding to the effect requested. This allows a single
  handling function to be installed in all relevant dynamic
  variables, but it is likely not the ideal choice.

dolio committed:
Add a test suite with some simple MCode.

- Includes various tweaks to ANF, MCode and interpreter to support some
  necessary operations, and fill in some missing cases that came up in
  actual snippets.

dolio committed:
Add missing MCode test file

- Also added some minor fixes/instances that make a test case work, and
  assist with debugging.

dolio committed:
Forgot to assert an MCode test.

dolio committed:
Implement a true lambda lifting pass on Terms

- First closes all lambdas with respect to free variables
- Second lifts all lambdas and let(rec)s to a single
  top-level binding group.

dolio committed:
Fix some problems with lambda lifting

- Let floating wasn't recursively floating bodies
- Non-recursive lets should only float if they directly bind
  lambda terms, and the lambda is what should be floated in
  that case.

dolio committed:
Various additions and fixes to properly test compilation

- Fixed benchmark with respect to new Reset form
- Made an ANF form for top level letrec on each term
- Changed lambda closure methodology
  * Instead of turning
      f = <v> -> e
    into
      f = (<w> <v> -> e) <w>
    change it into
      f = (<w> <v> -> e)
    and substitute `f <w>` in for `f` elsewhere
  * This has nicer properties with respect to subsequent
    transformations
- Don't float lets unless they immediately bind lambdas
- Don't float the lambda term out of its let binding
- Fixed a runtime bug with respect to argument preparation

dolio committed:
More minor fixes

- Ported argument preparation fix to unboxed arguments
- Fixed prompt handling for delimited control
  * It's not sufficient to track prompts by storing only the
    saved dynamic bindings, because there may be nothing bound
    in the associated variables.

dolio committed:
Introduce pattern compilation machinery.

- Turns match statements with nested patterns and
  guards into statements that only match on a single
  layer of a type, and do not use guards.

dolio committed:
Improve show instance for ABT

- Use precedence in Var case.

dolio committed:
Remove debugging statements from Unison.Runtime.Stack

dolio committed:
Various pattern compilation tweaks.

- Re-abstract variables for irrefutable rows
- Flatten @ patterns as part of an initial step
- Keep track of variable renamaings induced by pattern flattening
  and rename the variables in the body
- Syntactic cleanup

dolio committed:
Handle variable patterns in ANF

- Since pattern compliation happens before ANF, there is no
  good way to avoid binding variables in the pattern compiler,
  even though ANF only allows matching on something that has
  already been bound to a variable. So, instead, the ANF will
  take care of rebinding the variables.

dolio committed:
Add an MCode test case with compound patterns

dolio committed:
Define some builtins at the ANF level

- Some associated miscellaneous changes

dolio committed:
Support more ANF primops in the runtime

- Fix test/benchmark code for reworked runtime opcodes

dolio committed:
Modify MCode test to use builtins

- Tests break because the compilation is still unsuitable
- Also need to come up with new tests, because the builtins take up
  most of the hooks that were being used for running code. Should be
  feasible now that letrec is working, though.

dolio committed:
Various additions and reworkings to make tests work

- Moved Mem type to ANF module so that calling conventions can be
  tracked there
- Track calling convention in normalized syntax. Most surface level
  code compiles to boxed convention, while numeric literals and
  matching introduce some unboxed code.
- Resolve literals to their references for the above compilation
- Builtins use some unboxed operations as well. Optimizations may
  eventually move these around in the code.
- Added `Nat.sub` builtin
- Added n-ary unboxed calls to MCode
- Beefed up ANF->MCode compilation to make use of the calling
  convention tracking
- Emit direct unboxed matching for integer patterns since ANF now
  handles the indirection
- Fixed test denomralizer to handle new ANF transformations
- Reworked MCode tests now that builtins are being used for most
  reserved names.
- Show instances for ANF syntax

dolio committed:
Start implementing a primitive IO handler

dolio committed:
Delete a now misleading comment in the MCode generator

dolio committed:
Implement most IO functionality

- Add a notion of general foreign function that is used for most
  IO functionality. This is not especially fast, but most IO functions
  do not need to be.
- Add a type to enumerate the IO operations in ANF. Wrap them in
  code for builtins.
- Add a function that provides foreign function implementations of
  the ANF IO operations. Use it in emitting MCode.
- Wrapped, boxed values from Haskell have moved to their own module
  along with some other foreign implementation details. This allows
  unsafeCoerce imports to be limited to this module (although the
  foreign API provides the same unsafety).

dolio committed:
Add an Exit instruction that ends interpretation

- This can be used to handle a thread exit effect

dolio committed:
Add support for forking to the runtime

- Modify ANF for unboxed product/sum returns
- Add an ANF matching construct for discriminating on unboxed sums
- Add a Fork ending section and use it as the interpretation of the
  ANF fork primop
- Add an operation for duplicating the mutable components of the
  runtime stack
- Update tests to work with new code

dolio committed:
Add shift and continuation jumping to ANF language

- Shift is a built-in syntactic entity, used to compile handler
  matching. Previous compilation was incorrect.
- Jumping to a continuation is a new function call form. However,
  it is wrapped in a built-in because continuation and function
  values are distinct in the runtime.

dolio committed:
Switch fork operation to take a delayed computation

- This seems like a nicer building block, despite being more difficult
  to type (the sum version seems to require explicit delimited control
  to regain the expected semantics).

dolio committed:
Add thread kill/delay io operations

dolio committed:
Remove an unnecessary comment

dolio committed:
Flesh out foreign module with maybe calling convention

dolio committed:
Change emitPOp to have exhaustive matching

dolio committed:
Switch Unison.Runtime.MCode to explicit export

dolio committed:
Implement socket operations

dolio committed:
Eliminate some unused code

dolio committed:
Tweak foreign calls to allow catching exceptions

- Plumbing in unmasking of exceptions TODO

dolio committed:
Mask exceptions in runtime

- Foreign IO calls unmask to catch things

dolio committed:
Use unmasking action before forking

- Thread inherit original masking state, so I believe that not doing
  this would cause forked thread to never be unmasked if the parent was.

dolio committed:
IO calling conventions and misc.

- Error conditions from IO operations are now properly checked and
  translated back into the specified unison types
- Compilation of unboxed sum matching has been tweaked to support the
  incremental discovery of variables needed
- More IO operations have been implemented
- Some missing references were needed in IOSource

dolio committed:
Make a class for managing foreign input calling convention

- This obviates the need for a specific maybe function, for instance
- More compositional for compound types, as long as we don't try to
  pack information too much

dolio committed:
IOMode needs to be enum converted

dolio committed:
Fix some builtin foreign calling conventions

- Implement unenum for mashalling unison enums
- Implement translation for getBuffering

dolio committed:
Implement hSetBuffering compilation

dolio committed:
Implement Eq/Ord for closures.

dolio committed:
Remove committed commented code

dolio committed:
Various additions/refactorings in anticipation of decompilation

- Add wrappers for working with efficient maps/sets of Word64
  * HashMaps don't work for restricting a map to a set of keys
    efficiently
- Transition data type tags and references to be Word based. This
  doesn't affect anything about the references themselves, but the
  mapping to more space efficient values in the runtime
  * Data tags are now 16 bits; 65536 constructors should be plenty
  * Data declaration references may be 48 bits, so that they can
    be packed into the remaining space in a constructor tag
  * Function references can occupy the remaining space: 2^64-2^48
- Delimited control and dynamic scoping are now Word-based as well
- Created a map of known type references used in the builtin
  functions. It needs to be fleshed out with all the builtin
  references eventually.
- Data reference numbers are packed into data tags in the runtime,
  but masked off when dumping onto the stack so that case analysis
  remains on just the constructor.
- Deleted some dead test suite code

dolio committed:
Move tag packing stuff to where the tags are defined

dolio committed:
Start work on decompiling closures

dolio committed:
Extend decompilation to partial application

dolio committed:
Reference tagging related refactoring

- Term and type references don't need to share a numbering space,
  so they are now separated.
- Normalization code now looks up the relevant reference in separate
  ways.
- WordMap/Set has been generalized to arbitrary int-convertable key
  types. Unfortunately, the Enum class is inappropriate for this,
  because we want to just use the same bits for Word64, but Enum
  generates an error when the word is too large.
- Various ANF level structures have been given more precise types

dolio committed:
Move Rt2 to a more appropriately named location

dolio committed:
Fix imports/module header for Rt2 rename

dolio committed:
Fix miscellaneous errors in benchmark code

dolio committed:
Create interface to new runtime system

- Created a function for extracting references from LabeledDependency
  with an indication of whether it's a term or type reference
- Use a custom runtime exception type in the new runtime instead of
  erroring
- Create a new entry point to the runtime that applies a combinator
  * This function returns a closure result for watch expression use
- Use the pretty error type for decompile errors
- Names and backreferences for bultins
- Current runtime interface expects no actual referenced terms (just
  one large letrec). Some steps are in place for changing this, though.

dolio committed:
Merge branch 'master' into wip/rt2, with fixes

dolio committed:
Call into the new runtime in relevant places

- We may want to add a toggle later, but this allows experimenting
  directly with the new runtime.

dolio committed:
Fix ANF of letrec vars

- Top-level let rec variables cannot be used as function arguments,
  because they actually turn into references. So they must turn into
  a separate let in the ANF.

dolio committed:
Make it possible to show ANF code

dolio committed:
Change strategy for hooking into runtime

- The existing approach was wrong, because stack pointers are
  represented immutably. So, we need a hook to take the final
  machine state as its basis.
- Instead of just having the evaluation functions return a closure,
  we store an arbitrary hook as the final portion of the continuation.
  This allows us to read arbitrary stuff off the final stack.
- This is then used to write to an IORef that the runtime interface
  is expecting to be populated with the result in the case of watch
  expressions.

dolio committed:
Pretty printing for MCode

- Helps with debugging

dolio committed:
Fix compiling of final literals

- Originally the convention was to compile literals to boxed values,
  but that changed when ANF gained the capability to talk about
  unboxed things. However, only one spot of the translator was fixed
  to emit unboxed code for literals.

dolio committed:
Miscellaneous MCode improvements

- Forgot to export prettyComb in the relevant commit
- Primops were not being handled in the mcode compiler in one place
- Show the unhandled code when erroring in the mcode compiler
- Remove an outdated comment

dolio committed:
Add some primops to help with debugging

dolio committed:
Improve various error cases to give more information

dolio committed:
Handle boolean patterns in ANF compilation

dolio committed:
Add more global type references

dolio committed:
Handle booleans separately from other literals

dolio committed:
Another more informative error message

dolio committed:
Fix some pattern compilation behavior

- Implement an actual heuristic that will select relevant columns.
  Previously it was always selecting column 0, which was preventing
  elimination of some nested patterns.
- Re-abstracting over variables was incorrect. It was re-binding
  variables that had already been bound by higher splits, and was
  simply creating abstractions for final patterns that bind no
  variables. Instead, we should bind any non-trivial term-variable
  pairs with a let binding.
- The correct handling of remaining term/variable rows revealed a
  mismatch in how pattern matrices are split. When a row was split
  on a variable, it was keeping a term but eliminating the column,
  causing there to be more terms than patterns.

dolio committed:
Handle variable bindings in guards

- Currently assumes that the bound variables are identical to the
  bound variables in the right hand side. This seems like a safe
  assumption, but if we see an error because it isn't, a more
  elaborate approach can be implemented.

dolio committed:
Support for more literals; text and double

dolio committed:
More information in decompilation errors.

dolio committed:
Implement most floating point primops

dolio committed:
Implement universal equaltiy

- The underlying Eq code may need some tweaking in the future

dolio committed:
Add Float.atan2 and Text.empty

dolio committed:
Rework Foreign to support better universal equality

- Foreign now stores a unison reference to indicate the stored type.
  This allows implementing equality on what would otherwise be an
  opaque box.
- Unboxed foreign arguments just get sent as int, since there is no
  way to reconstruct the correct reference dynamically.

dolio committed:
Fix code emission for text literals

dolio committed:
Provide more information in a pattern compiler error

dolio committed:
Implement various text operations

- Add boxed primitive types, which now includes universal equality
- Various convenience functions as well

dolio committed:
More primop work

- Eliminate some less than instructions, since they can be
  done without
- Implement some bitwise operations
- Integer casts via a coercion builtin
- Fix Int parsing to allow a leading '+'

dolio committed:
Initialize closure arrays to BlackHole instead of undefined

dolio committed:
Implement text pattern matching

dolio committed:
Include some missing builtin type refs in the global map

dolio committed:
Support decompiling text

dolio committed:
New calling/return convention for variable arguments

- New convention is DArgV (v for 'variable') for yielding results
  from a primop that doesn't have a fixed return shape.
- Returns all values past a specified offsets in the frame. The
  length of the 'arguments' being passed is determined dynamically
  based on the pointers.
- Also fixes some bugs with the ArgR calling convention, since the
  variable argument return is implemented that way dynamically.
- This allows all functions ending in a primop call to be implemented
  with the variable return convention, rather than having to emit
  the return convention of every primop (which isn't even possible in
  general).
- Previously the convention was assumed to be 1 unboxed value, which
  is no longer correct.

dolio committed:
DArgV was counting too many variables in a Let

- Let can introduce a context where a Yield should not eliminate
  all the variables in the frame. So we need to keep track of which
  variables were introduced by the Let block.
- This was the breaking point for introducing a more structured
  variable context type.

dolio committed:
Additional error message information

- Decompiler and primop compilation

dolio committed:
Fix some builtin bugs

- Numeric cast was leaving the type tag in place
- Nat.drop had its arguments backwards
- Nat.sub was returning Nat instead of Int

dolio committed:
Int.truncate builtin was not tagging correctly

dolio committed:
Rework pattern compilation significantly

- Avoid having to figure out when we need to rebind variables for
  VarP cases by pre-binding scrutinees and only casing on variables.
- Make use of constructor information for compiling data, so that
  cases are filled out based on all constructors. This fixes previous
  incorrect behavior with respect to how variable/default cases were
  being handled
- Built-in special patterns work similarly to the old methodology, but
  don't have the complications of data matching (aside from sequences,
  which are still TBD).

dolio committed:
Don't try to determine field references from constructor type

- It doesn't work very well with the data declaration information
  that is tracked by unison
- Also include a previously uncommitted change from data declarations,
  which now gives the number of fields that a constructor has (this is
  the only information the pattern compiler now relies on; previously
  it was assuming the Reference for the type of each field could be
  determined from the DataDecl).

dolio committed:
Use UnboundP for defaults in builtin matching

- This was the purpose of ensuring we only match on variables,
  and builtin matches are the only ones that have default
  cases anymore (at the moment)

dolio committed:
Keep track of effect vs. data in pattern compilation

- We need to rebuild the right patterns, and effect declarations
  don't actually have the 'pure' case in them (which is uniform
  across all effects), so they need slightly different handling.
- Also duplicate a bit more 'builtin' logic to avoid threading
  around Either values where only one branch makes sense in any
  situation

dolio committed:
Strip annotations before lambda lifting

- It was causing some odd behavior during the lifting, and the
  annotations aren't necessary at that point.

dolio committed:
Prepare UnboundP as VarP once more

- This was a mistaken attempt to fix the default case in some
  situations, but it has been fixed in other ways now, and interacts
  inappropriately with some other pattern compilation code.

dolio committed:
Eta expand constructors before floating

dolio committed:
letFloater was throwing away context from sub-terms

dolio committed:
Arrange for handler bodies to be floated

- There is currently a bit of indirection. Handle blocks delay
  evaluation so that the delayed value can be passed to a handler
  setup function. So, any non-trivial computation has to be floated
  into a form that can be delayed.

dolio committed:
Implement some pretty printing of ANF terms for debugging

dolio committed:
Add TFrc pattern synonym

- Was missing since AFrc was added
- Updated pattern completeness pragmas

dolio committed:
ANF pretty printing tweaks

- Take lambda calling conventions into account so we can detect if we
  have a mismatch between variables and conventions.

dolio committed:
Handler compilation was missing a calling convention

dolio committed:
Fix calling functions with top-level arguments

- Top-level functions are referenced global numbers, but functions must
  be called with Closures. If we want to pass a top function as an
  argument, we must build a closure referencing it.

dolio committed:
Fix pattern compilation base case

- Previously it was erroring on an empty matrix, but it is possible
  to arrive there via guards. So, instead, emit an erroring term in
  that scenario, rather than failing compilation.

dolio committed:
Implement term/type links

dolio committed:
Implement various sequence builtins

dolio committed:
Finish sequence operations and literals

dolio committed:
Nicer argument printing in combinators

dolio committed:
Add missing some builtin type references

dolio committed:
Decompile characters

dolio committed:
Show arguments when pretty printing combinators

dolio committed:
More ANF pretty printing tweaks

dolio committed:
Fix Yield to handle e.g. `id id 5`

- When yielding a result from a function, pending arguments were
  being ignored. Instead, we should see if we're yielding something
  eligible to be applied to the pending arguments, and do so.
- There is currently no special behavior for situations where
  something else is being yielded. In principle this is a type
  error.

dolio committed:
Make sure to pack pending arguments in augSeg

- This fixes a scenario where over-application of a wrapper that
  results in a partial application of the wrapped function would
  lose arguments.

dolio committed:
Fix partial application's argument handling

- The pending arguments were correctly getting packed into the closure,
  but that does not discard them from the stack, and the `apply`
  function was neglecting to do so. So certain arguments could be left
  on the stack in some situations, messing up the indexing.

dolio committed:
Fix a variable duplication/catpure problem in floating

dolio committed:
Name and pending argument tweaks

- Name now takes a reference, so we may delay evaluation of a
  call to a stack argument.
- augSeg now takes a parameter indicating whether it should consume
  pending arguments. It should not in the case of Name and Pack for
  instance.
- augSeg also takes a Maybe Args', because it may be necessary to
  call it in a situation where only pending arguments are available.
- closeArgs takes the pending argument configuration to pass to augSeg

dolio committed:
Fix code emission for requests

dolio committed:
Fix segment agumentation for continuations

- The strategy for jumping to a continuation is to pack arguments
  onto the captured stack, discard the current frame, then dump the
  segment. However, the previous code was augmenting stacks in a way
  that assumed the segment was captured arguments, which would go on
  the opposite end of the segment than a captured stack.
- Replaced the boolean flag of augSeg with a mode type that specifies
  which of the three modes it should run in. Instruction mode does not
  consume pending arguments, call mode does, and continuation mode
  packs the arguments on the other end of the segment.

dolio committed:
Fix group references for lifted handler functions

- letrec variables were being counted as free variables for functions
  lifted out of handler matching, so the straight-forward handler
  function would refer to itself indirectly.
- I think it may be possible to create a shadowing scenario where this
  causes incorrect behavior. Some investigation is required.

dolio committed:
Add fake boolean data info to pattern compiler

dolio committed:
Fix overapplication of ability requests

- The code emitter assumes that requests are exactly saturated, similar
  to constructors. However, unlike constructors, requests can be over
  applied. The current calling convention (where the emitter packs
  arguments in a data type) prevents the runtime from fixing things up,
  so instead add a pass to enforce exact saturation of requests.
- Thus, constructorEta has been renamed to saturate

dolio committed:
Modify heuristic for floating handle bodies

- Constructors don't need to be floated to a named definition, even if
  not nullary. We must allow them later because literals turn into
  unary constructors.
- Applications of a function variable are structured. Only constructor
  applications may be considered unstructured.
- Constructor applications may be locally floated out of a handle during
  ANF, because they don't incur any effects.

dolio committed:
Derive show for CTE

dolio committed:
Add some debugging helpers for the runtime

dolio committed:
Implement universal comparison operators

dolio committed:
Handle sequence comparison

dolio committed:
Fix off by 1 error in PApV implementation

dolio committed:
Fix variable duplication/capture in floating

dolio committed:
Show reference in show for Foreign

dolio committed:
New variable handling in float was missing some variables

dolio committed:
Remove some debugging that snuck into a previous commit

dolio committed:
Implement bug/todo

- Nice error messages pending

dolio committed:
Implement sequence pattern compilation

- The strategy used for sequences is fairly different, because
  the patterns do not actually correspond to any kind of essential
  structure. So compilation of complicated patterns involving sequences
  may end up performing multiple operations on scrutinized values.
- The compiled sequence patterns are not yet handled downstream

dolio committed:
Handle sequence patterns in compilation/machine

dolio committed:
Miscellaneous runtime interface fixes

- Missed committing proper handling of runtime exceptions
- Combinator numbering was not happening appropriately for all
  decompilation

dolio committed:
Refinements of sequence pattern compilation

- It is still not completely correct, but dealing with the problem fully
  requires a modified approach with more lax 'matrices'.

dolio committed:
Missed adding Unison.Runtime.Debug to cabal file

dolio committed:
Merge branch 'trunk' into wip/rt2, with fixes

- Fixed some problems with previous commits due to local build failure

dolio committed:
Rework pattern matching machinery; finish sequences

- Pattern 'matrices' no longer need to be rectangular. Instead they
  just track which refutable patterns are necessary for each row,
  tagged with the variable to match against it.
- Make splitting matches have a split and non-split case for more
  uniform handling
- Change freshening strategy for making up pattern variables

dolio committed:
Splitting primops were testing backwards

dolio committed:
Add toggles for new runtime, with default to existing runtime

- There is both a config file setting and a command line
  parameter, with the latter overriding the former.

aryairani commented:
Awesome!

> The majority of the test suite passes using the new runtime. One failure is an actual incorrect test case (#1613). The other is the recently added IO case, which I haven't looked into just yet. There are also some transcript differences that I haven't resolved yet.

IMO, we should address these before merging, and have CI run the tests on both the old runtime and the new one (if that's not already part of this PR).

> In general, the IO implementation is relatively untested, and may have bugs, but I thought this pull request might signal that someone besides me could actually try things out, as I might not find all the issues myself.

Totally.  I think once existing tests passing under CI, we merge it to master and let people start trying it with the flag.  (Though they are welcome to try it in this branch?) Once an arbitrarily short amount of time passes, we can switch the default to the new runtime; and once the old runtime is known to be strictly worse or at least buggier in a way that we don't want to fix, we can dump it altogether.

atacratic commented:
Awesome this is landing!

Is it for definite the destination is bin the old runtime, rather than keep both? The old one might act as a simple reference implementation, and (I'm thinking at some point) one with extra instrumentation for analysing the code being run. Instrumenting the new runtime might be more of a challenge since it's more involved, and also the priority will be avoiding compromising speed.

aryairani commented:
@atacratic I think we can keep both as long as they both work; I was just assuming that over time we'll encounter issues with the old one that nobody wants to fix, and under those circumstances, it'd probably not be worth keeping.

atacratic commented:
Is there a way to rerun all the tests in a codebase using the new runtime? Without the test cache preventing it?

atacratic commented:
@dolio I brought the new runtime up in my codebase and ran 'test' and hit the following.  Known issue?  LMK if you want more info.

```
chris@chris-thinkpad:~/unison$ stack exec unison -- --new-runtime

   _____     _             
  |  |  |___|_|___ ___ ___ 
  |  |  |   | |_ -| . |   |
  |_____|_|_|_|___|___|_|_|
  
  Welcome to Unison!
  
  You are running version: devel/M1l-189-gdafb3c57
  
  I'm currently watching for changes to .u files under ~/unison
  
  Type help to get help. 😎


  I loaded scratch.u and didn't find anything.

.> names base.List.Nonempty.foldr.test

  Term
  Hash:   #07s0393c3l
  Names:  _base.releases._M1m.List.Nonempty.foldr.test
          _base.releases._latest.List.Nonempty.foldr.test
          _base.series._M1m.List.Nonempty.foldr.test
          _base.trunk.List.Nonempty.foldr.test
          base.List.Nonempty.foldr.test

.> cd base.List.Nonempty
.base.List.Nonempty> test

    
    Cached test results (`help testcache` to learn more)
    
    ◉ scanl.test    : Passed 100 tests.
    ◉ scanr.test    : Passed 100 tests.
    
    ✅ 2 test(s) passing
    
    ✅  

14 tests left to run, current test: foldr.testunison: buildMatrix: empty rows
CallStack (from HasCallStack):
  error, called at src/Unison/Runtime/Pattern.hs:275:18 in unison-parser-typechecker-0.1-AVd5UsYOj5GOFXnV4xcly:Unison.Runtime.Pattern

```

dolio commented:
No, not a known issue. I've only been running the build time test suite, though, not any tests written inside unison.

pchiusano commented:
NICE!! Looking forward to this. 

@dolio I would work on getting all the tests in base to pass with the new runtime, and we should turn that into a a test transcript at some point (maybe run separately from the other transcripts since it will fetch base):

```
.> pull https://github.com/unisonweb/base:.releases._latest base
.> cache.clear -- doesn't exist yet, but we should add a command for this
.base> test
```

I'd like to see some high level documentation of the compilation process. Can be pretty short I think. Basically, what are all the types involved, what are the phases of compilation, where does the code for different phases live and what are interesting things along the way. We should do a synchronous review and come up w/ rough outline for this once you've ironed out the bugs you find in getting all the base tests to pass. 

I'd also like to see comments in the code for all types and the constructors if they aren't self-explanatory. Like there's stuff in `MCode` with some abbreviations whose meaning isn't obvious, and probably other places as well.

https://github.com/unisonweb/unison/blob/af12fab03f2638fdd2588d5e4ae3a6bb37bbac5b/parser-typechecker/src/Unison/Runtime/MCode.hs#L360

dolio commented:
Yeah, I've been neglecting some documentation more than I like as time went on, because I've just been trying to get the implementation done. You can see that there's a big writeup of the runtime ideas from a while ago, but less about newer stuff. I'll try to fill it in. The `MCode` instructions almost all have analogues in the `ANF` code that do say what they are, but it doesn't hurt to say in both places.

dolio committed:
Runtime interface was failing to load some information

- The builtins need to refer to some declared data types, but we don't
  populate the runtime context with some of the information for those.
  So the check for not loading the information has been relaxed
  somewhat.

dolio commented:
So, I've been digging more into the IO stuff, and there's a pretty significant problem, I think. The `IOSource` stuff is written in a way that is pretty specific to the old runtime choices. For instance, it thinks file handles are text. So, that carries over to the code it feeds into the new runtime. But, file handles in the new runtime are expected to actually be file handles.

I kind of think that some of these choices don't really make sense anyway. For instance, treating file handles as text keys into a local map is just giving a false sense that they can be stored or transferred to other machines. However, I don't know whether it's worth it to modify the existing runtime at the moment.

pchiusano commented:
I wouldn't worry about updating the old runtime or trying to have new stuff match what IOSource is currently doing.

dolio commented:
Well, the point is that this is no longer just swapping out a runtime implementation. There's a big embedded text file that no longer entirely makes sense, and it seems way higher in the pipeline than the runtime. Also, there are definitions in it that are necessary for the new runtime, like `EIther`. The references to those definitions are named as Haskell values, and used in the primops to build the right data and such. I also don't know if it makes sense to swap that text file out from under a repository that builds on the existing one.

dolio committed:
Add builtins for new IO functions

- The ability for this is a built in reference, rather than a declared
  definition, so it won't be possible to handle it without some
  additional work. The plan is for it not to be possible to handle it,
  though.
- Functions using the ability are directly implemented, rather than
  going through indirections.

dolio committed:
Preprocess top let recs for proper running

- When running an IO action, we get a recursive let like:

    let rec
      x = 'action
      x1 = !x
    in ()

  The intention of this is for the () evaluation to incur the effect of
  x1. However, let recs are assumed to just be introducing definitions,
  not evaluating things. So, instead we pre-process the top letrec,
  peeling off pre and post let blocks that can evaluate things. The
  middle letrec should eventually be floated above these.

dolio committed:
Assorted new runtime IO changes/fixes

- Adjusted calling conventions to match the fact that there are no
  data wrappers around most types; they are effectively foreign imports.
- Removed the `handle'io` definition, as the intention is for the new IO
  interface to be wired in and implemented directly, not a handler for
  some defined ability.
  * Such an ability (or many fine grained abilities) could be defined
    and handled into the builtin definitions.

dolio committed:
Tie main type to chosen runtime implementation

dolio committed:
Add a runtime debugging function for single Combs

dolio committed:
Change strategy for running main entry points

- The letrec processing function for the new runtime was overly naive,
  and would have required something like topological sorting to make
  work well.
- Instead, we add the forced main entry point to the watches of the
  unison file generated for running. This causes the forcing to go in a
  normal let, which will ensure the runtime evalutes it, rather than
  just defining it.

dolio committed:
Make buildMatrix behave better on zero rows

- Previously this was something of a corner case, but in light of some
  of the rewrites, it seems sensible to just build an empty matrix in
  all cases, and this obviates special cases elsewhere in the pattern
  compiler.

dolio committed:
Add builtin or/and functions to the new runtime.

- The compiler to the new runtime references these in some situations

dolio committed:
Change float variable naming for better debug printing

- When floated, local names need to be freshened for when they appear in
  a single scope. Internally the freshening is done by attaching an
  integer to the named variable. However, the pretty printer assumes that
  names freshened in this way do not need to be displayed distinctly. So
  the floating pass creates genuinely newly named variables that are
  fresh in the term.

dolio committed:
Change ANF freshening strategy

- Rather than trying to keep track of things to avoid, just store a
  known word to freshen by. All variable freshening in ANF is for new
  variables, not freshening of existing variables.

dolio committed:
Implement Debug.watch

- Repurposed the print instruction for this.

dolio committed:
Properly handle shadowing binding group variables

- There are situations where we must build a closure for a letrec
  definition, but the previous handling of that was assuming that the
  letrec bindings were never shadowed. This was handled in the ANF code,
  and resulted in bad code being generated further down the line.

dolio committed:
Improve choice of common variables in pattern compilation

- The previous code was simply looking at the first row to choose a
  common set of variables to use for the pattern. However, the first row
  could have been generated from an irrefutable pattern, which binds no
  subpattern variables. So, it is more robust to look for a row that
  actually has non-trivial subpatterns.

dolio committed:
Improve variable closure logic

- Previously, the logic was simply replacing e.g. `f` with
  `f x y z` where `x` `y` and `z` were free in the body of `f`. However,
  this does not account for the fact that doing so may add additional
  free variables to the bodies in question. So, a more complex tracing
  procedure was needed to determine all the variables needed to make a
  self-contained combinator.

dolio committed:
Implement universal equality in terms of compare for now

- Universal compare was designed for better extensibility, and can
  support sequences. A similar equality can be built if necessary.

aryairani committed:
Merge branch 'topic/patternp' into wip/rt2

dolio committed:
Various fixes for in-unison tests

- Bumped the dummy stack ensurance for combinators to 20, because some
  functions were using more than 10 framed values. Actual proper
  calculation still TODO
- Made sequence pattern matching call a boxed function instead of the
  direct unboxed operation. The unboxed return causes problems when it
  happens in the middle of a function, because it is unclear how many
  things are on the stack afterwards. A proper fix seemed unforthcoming,
  so boxing is simpler in the mean time.
- Added string (un)packing builtins.

dolio committed:
Change strategy for handler pure cases

- Previously the pure case was inserted into the code block under the
  reset. However, this causes it to be part of the continuation captured
  by any shifting, which causes duplication of pure cases when more than
  one effect is used.
- Instead it has been moved into the actual handler function. The pure
  case is treated as the RTag 0 case, while other abilities necessarily
  have positive RTags.
- When yielding past a mark on the continuation, we call into the
  handler with this case.
- Various other code has been shuffled a bit to make all this happen.
- This seems to have fixed the remaining base test cases.

dolio committed:
Allow nullary application of non PAps

- This allows values to serve as their own 'lazy' binding
  representation, and forcing them to yield correct results.

dolio committed:
Implement necessary frame size calculations

- Previously we were just conservatively estimating how big a frame
  a function could require.

dolio committed:
Fix test compilation for previously modified handler adjustment

dolio committed:
Implement Bytes builtins

dolio committed:
Commit transcript changes for new builtins

- These were generated using the existing runtime

dolio committed:
Merge pull request #1649 from unisonweb/wip/rt2-patternp

merge topic/patternp into wip/rt2

dolio committed:
Fix test discrepancies

dolio committed:
Merge branch 'trunk' into wip/rt2

dolio committed:
Unison.Runtime.MCode documentation

pchiusano commented on parser-typechecker/src/Unison/Builtin/Decls.hs:
```diff
@@ -115,6 +129,68 @@ builtinDataDecls = rs1 ++ rs
+      )
+    ]
+  fmode = DataDeclaration
+    Structural
```
@dolio any of the file / IO enums should be `unique` types. You pick the GUID, via `head -c 1024 /dev/urandom | shasum`.

dolio replied:
Why, though? Is this because a structural definition can only have one set of relevant namings?

pchiusano replied:
Well, a same structural type can have multiple names, but it's more that without `unique`, the type will be considered the same as like `type Four = One | Two | Three | Four` and they could pass a `Four` where a `FileMode` was expected. Since `FileMode` has a specific meaning that's probably not what we want. (vs say `type List a = Nil | Cons a (List a)` where the structure is really all there is to say about the type) 

Another example is like `unique type Day = Mon | Tue | ...` which is probably not something you want to be confused with "something with 7 nullary constructors" :)

Aside: we are likely going to make it a type error or a very stern warning to have a structural type where two of the constructors have the same type. /cc @aryairani

dolio replied:
Here's another question: other builtin unique decls have different length strings. Some look like the length of `sha1sum` and others are probably like `sha256sum` or larger. Does it not matter?

aryairani replied:
It doesn't matter for now; currently they're just strings that are parsed as `wordyId`s. 😅

pchiusano commented on parser-typechecker/src/Unison/Builtin/Decls.hs:
```diff
@@ -115,6 +129,68 @@ builtinDataDecls = rs1 ++ rs
+    , ((), v "FileMode.ReadWrite", var "FileMode")
+    ]
+  bmode = DataDeclaration
+    Structural
```
same deal here, should be `unique`

dolio replied:
These are now done.

pchiusano commented on parser-typechecker/src/Unison/Builtin/Decls.hs:
```diff
@@ -115,6 +129,68 @@ builtinDataDecls = rs1 ++ rs
+      , Type.nat () `arr` var "BufferMode")
+    ]
+  ioerr = DataDeclaration
+    Structural
```
`unique` here too

pchiusano commented on parser-typechecker/src/Unison/Builtin/Decls.hs:
```diff
@@ -115,6 +129,68 @@ builtinDataDecls = rs1 ++ rs
+    , ((), v "PermissionDenied", var "IOError")
+    , ((), v "UserError", var "IOError")
+    ]
+  seqview = DataDeclaration
```
Can you explain how this is used? Just curious.

I think it'd be (in Unison syntax) `SeqView a = Empty | SeqView a (SeqView a)` which is just a cons list, right?

I could also imagine `SeqView a seq = Empty | SeqView a seq` which you might use for uncons-ing or unsnoc-ing "one step at a time".

dolio replied:
Yeah, this was a mistake. It should be more like you say.

dolio replied:
Okay, I made this basically `type SeqView a b = Empty | Elem a b`. This fixes something else that was bugging me too, because it gets used for both left and right viewing, so either side may be the sequence depending on which you do.

pchiusano replied:
Ah, nice! It's always bugged me that there's a separate `ViewL` and `ViewR` type in Haskell for like Data.Sequence.

pchiusano commented on unison-src/transcripts/alias-many.output.md:
```diff
@@ -18,174 +18,251 @@ Let's try it!
+  75.  Float.truncate : Float -> Int
+  76.  IO.clientSocket : Text
+                         -> Text
+                         ->{##IO} Either IOError ##Socket
```
This transcript calls `builtins.merge` at the start, which looks like it includes names for the builtin `IO` functions but not for the `IO` type itself? (since it is showing up as `##IO` here). That's a little odd.

We could use `builtins.mergeio` to give names for both the `IO` type and all these functions. 

Or just have `builtins.merge` include the `IO` type and functions.

@aryairani WDYT?

aryairani replied:
Side note, based solely on transcripts from a while back, it seemed like the IO builtins were being put a level too high.  And yeah we should check for an extra `#`.

dolio replied:
The old `IO` ability still exists (and is used in the existing runtime), so any ideas on what the names should be?

aryairani replied:
Ah, yeah if they both have to coexist, I guess we could call it `IO2` and rename it later.  The text would remain `IO2.whatever` even when we give it a new name in the namespace.

pchiusano replied:
The builtin names can just be `IO.whatever`, but I'd just put them all in an `io2` namespace. This could be done with rename directives or in the spot where we `builtins.merge[io]`, wdyt @aryairani ?

The existing `IO` type and functions aren't builtins, so there's no name conflict there.

aryairani replied:
Oh I see, they aren't builtins, right.  So then there's no rename directives (those only applies to Builtin types and terms), but yeah we can something something them in an `io2` namespace in `builtins.mergeio(2?)`).

pchiusano replied:
@aryairani Cool :) Can you help @dolio with making this change?

I was imagining adding like `R "IO" "io2.IO"` to the little DSL in `Builtin.hs` but I think that is a bad idea and better to just change where it is merged into the namespace tree like you suggest.

dolio replied:
Everything related to the new `IO` stuff is now under the `io2` namespace when you merge it, I believe. I made the declared types go there, too, for instance. I presume just giving them that name in the hand-built declaration works, since there's no renaming DSL like there is for the builtin types.

pchiusano reviewed (changes requested):
Hey @dolio awesome that this is finally getting in!! Am excited about all the cool stuff we are going to be able to start adding.

I think the one 'must have' before merging is switching those various enum types to `unique` instead of structural. I left a few other minor comments.

dolio committed:
More documentation

dolio committed:
Refine some aspects of the new runtime

- Gave names to some of the new builtin types involved in I/O during
  builtin merge
- Switched the new IO enums to be unique types
- Moved new IO related stuff to be in the io2 sub namespace on builtin
  merge

pchiusano reviewed (approved):
Okay, I think this is good to merge! @aryairani can you review how @dolio did the `io2` stuff. I think it should be fine, but I'd like your sign off. You can merge if it looks good and no other blocking issues.

pchiusano commented:
Also @dolio can you check the CI failures and see what's going on there? I feel like these were all green at one point previously.

dolio commented:
Yeah. I think I just forgot to redo the transcripts with the new names.

aryairani commented:
> @aryairani can you review how @dolio did the `io2` stuff.

Will do.

dolio committed:
Fix transcripts after IO naming fixups

dolio committed:
Make IO test disambiguate which IO it's using.
